### PR TITLE
Make use of default basemap for new Builder maps

### LINKED ISSUE #12700: Add Voyager style to CARTO basemap selection WAITING FOR DEPLOY DATE
Technical info (URLs) is given here: https://github.com/CartoDB/website/issues/3953

### PR DESCRIPTION
**Pull request by juanignaciosl**
Closes the code side of #12700, and #12810. Keep in mind that the second one is a refactor with high impact. Check issue casuistry for acceptance as well.

In production, we must ping @noguerol before doing the configuration change, because he'll coordinate some communication efforts.

Deploy instructions (take this into account for Acceptance):
- [ ] Deploy this PR. Everything should work as it does now.
- [ ] Change configuration, adding Voyager basemap.

### CODE REVIEW

**Comment by juanignaciosl** [comment]
retest this please

File: config/app_config.yml.sample [hunk 1]
```diff
@@ -378,10 +378,35 @@ defaults: &defaults
   datasources:
     arcgis_enabled: false
     salesforce_enabled: false
-  basemaps:
+  basemaps: # First one becomes default one
```
Comment (javitonino) [inline config/app_config.yml.sample]: Why this change instead of using the `default: true` attribute?

File: config/initializers/01_app_config.rb [hunk 2]
```diff
@@ -65,6 +65,10 @@ def self.asset_path
     end
   end
 
+  def self.default_basemap
+    Cartodb.config[:basemaps].values.first.values.first
+  end
```
Comment (javitonino) [inline config/initializers/01_app_config.rb]: Not a fan of adding a global method only used in tests. For tests, I would favour direct comparison against a fixed string (which can be a global constants, but maybe in a spec helper).

File: config/app_config.yml.sample (hunk 1 above)
Comment (juanignaciosl) [reply config/app_config.yml.sample]: The default was not working. Take a look, just in case I was wrong, but I think that we're just getting the first one. Instead of changing the behaviour I opted to delete the (wrong) documentation.

File: config/initializers/01_app_config.rb (hunk 2 above)
Comment (juanignaciosl) [reply config/initializers/01_app_config.rb]: Meh. The problem is that a configuration change might lead to a test breaking, which seems too fragile. I did this method to abstract the "default basemap" decision. I would use it at production code if needed, indeed.

In case that you don't want this, I'd inline it instead of using strings, though.

Opinion?

File: config/app_config.yml.sample (hunk 1 above)
Comment (javitonino) [reply config/app_config.yml.sample]: It seemed like it was using it, at least it is reading the `default` attribute. https://github.com/CartoDB/cartodb/blob/master/app/models/carto/user.rb#L234-L241
But it is only using it to find the default group (i.e: CARTO, Stamen...), and then takes the first basemap there. Inside the group, it takes the first basemap.

So `default` gives us something, even though it makes little sense to put the default inside one specific basemap, instead of the group. So I think this change is ok, and something maybe we could take a deeper look at some other time. 👍

File: config/initializers/01_app_config.rb (hunk 2 above)
Comment (javitonino) [reply config/initializers/01_app_config.rb]: Well, this is also kind of fragile. Taking from the previous comment, if I removed `default: true` from CARTO and added it to `Stamen`, this would fail.

Anyway, my concern is having some production code with the same name that the method in the user model, which does almost the same thing, but not really. This one does not implement the default per group, while the one at user model does.

I'm okay with this method living in test code, but if it is going to be in production, it should return exactly the same thing as the user model (probably, the user model one should call this for non-google cases).

Comment (juanignaciosl) [reply config/initializers/01_app_config.rb]: Ok, back to development then, I'll take a look.

File: config/app_config.yml.sample (hunk 1 above)
Comment (juanignaciosl) [reply config/app_config.yml.sample]: I've changed the comment to be more accurate, although this should be improved.

**Comment by javitonino** [comment]
🍏 acceptance
Builder always set up default basemap from config
Editor always keep basemap from dataset

File: config/app_config.yml.sample [hunk 3]
```diff
@@ -378,10 +378,36 @@ defaults: &defaults
   datasources:
     arcgis_enabled: false
     salesforce_enabled: false
-  basemaps:
+  basemaps: # Adding `default: true` at a basemap marks its group as the default one. Its first basemap becomes the default one.
     CARTO:
-      positron_rainbow_labels:
+      voyager_labels:
         default: true
+        url: 'http://cartodb-basemaps-{s}.global.ssl.fastly.net/rastertiles/voyager_nolabels/{z}/{x}/{y}.png'
```
Comment (jaakla) [inline config/app_config.yml.sample]: Perhaps use here `http://{s}.basemaps.cartocdn.com` domain just as in other basemaps? Only advantage of CDN-provider domain _...fastly.net_ is that it works also with https, but if here only http is used, then cartocdn.com is better as it does not have specific 3rd party cdn domain in it.

Comment (jaakla) [reply config/app_config.yml.sample]: @javitonino  - this reminds me another thing , shouldn't the basemaps request use same security as HTML - i.e. https ? I do not know if it is functional issue really, but browsers give mixed content warnings at least to developers, and I've heard that in long term chrome plans to deprecate http, so better opt everything what we can to https? I'm trying to get https for main own domain also (https://github.com/CartoDB/lbs-services/issues/3), so far fastly.net can be used for https, and there is no extra cost for it.

Comment (jaakla) [reply config/app_config.yml.sample]: I did some tests, for http/https autoselection following method seems to work fine in js/leaflet level at least:

      layer = L.tileLayer('//cartodb-basemaps-{s}.global.ssl.fastly.net/'+style+'/{z}/{x}/{y}' + (L.Browser.retina ? '@2x.png' : '.png'), {

Comment (juanignaciosl) [reply config/app_config.yml.sample]: I used the Fastly domain schema because currently that's how most basemaps are configured. Example: `'https://cartodb-basemaps-{s}.global.ssl.fastly.net/light_nolabels/{z}/{x}/{y}.png'`. While `eco_cartodb`, `flat_blue`, `midnight_cartodb` and `antique_cartodb` use `http://{s}.basemaps.cartocdn.com`, others don't. I don't know if there's an specific reason for that.

But it's true that it should be https, not http.
